Data/ByteString/Lazy.hs

```diff
@@ -696,6 +726,55 @@ drop i cs0 = drop' i cs0
             then Chunk (S.drop (fromIntegral n) c) cs
             else drop' (n - fromIntegral (S.length c)) cs
 
+-- | /O(n)/ @'dropEnd' n xs@ is equivalent to @'take' ('length' xs - n) xs@.
+-- Drops @n@ elements from end of bytestring.
+--
+-- >>> dropEnd 3 "abcdefg"
+-- "abcd"
+-- >>> dropEnd 0 "abcdefg"
+-- "abcdefg"
+-- >>> dropEnd 4 "abc"
+-- ""
+--
+-- @since 0.11.2.0
+dropEnd :: Int64 -> ByteString -> ByteString
+dropEnd i p | i <= 0 = p
+dropEnd i p = go D.empty p
+  where go deque (Chunk c cs)
+            | D.elemLength deque < i = go (D.snoc c deque) cs
+            | otherwise          =
+                  let (output, deque') = getOutput [] (D.snoc c deque)
+                    in L.foldl (flip chunk) (go deque' cs) output
+        go deque Empty           = dropElements deque (fromIntegral i)
+
+        len c = fromIntegral (S.length c)
+
+        -- get all `S.ByteString` from the front of the accumulating deque
+        -- for which we know they won't be dropped
+        getOutput out deque =
+            case D.popFront deque of
+                 (Nothing, deque') -> (out, deque')
+                 (Just x, deque')  ->
+                    if len x <= D.elemLength deque' - i
+                       then getOutput (x:out) deque'
+                       else (out, deque)
+
+        -- drop n elements from the rear of the accumulating `deque`
+        dropElements deque n
+            | D.null deque = Empty
```

Comment:
I think you do not need to handle this case specially: if `deque` is empty, then `D.popRear` returns `Nothing` and `Empty` naturally arises below.